tracer/src/Datadog.Trace/AppSec/Coordinator/SecurityCoordinator.Framework.cs

```diff
@@ -69,23 +78,86 @@ internal SecurityCoordinator(Security security, HttpContext context, Span span)
     internal IDictionary<string, object> GetPathParams() => _context.Request.RequestContext.RouteData.Values.ToDictionary(c => c.Key, c => c.Value);
 
     /// <summary>
-    /// Initializes a new instance of the <see cref="Coordinator.SecurityCoordinator"/> class.
-    /// framework can do it all at once, but framework only unfortunately
+    /// Framework can do it all at once, but framework only unfortunately
     /// </summary>
     internal void CheckAndBlock(Dictionary<string, object> args)
     {
         var result = RunWaf(args);
         if (result.ShouldBeReported)
         {
+            var reporting = MakeReportingFunction(result.Data, result.AggregatedTotalRuntime, result.AggregatedTotalRuntimeWithBindings);
+
             var blocked = false;
             if (result.ShouldBlock)
             {
-                blocked = WriteAndEndResponse();
-                _httpTransport.MarkBlocked();
+                blocked = ChooseBlockingMethodAndBlock(reporting);
+            }
+
+            reporting(blocked);
+        }
+    }
+
+    private Action<bool> MakeReportingFunction(string triggerData, ulong aggregatedTotalRuntime, ulong aggregatedTotalRuntimeWithBindings)
+    {
+        var securityCoordinator = this;
+        return blocked =>
+        {
+            if (blocked)
+            {
+                securityCoordinator._httpTransport.MarkBlocked();
+            }
+
+            securityCoordinator.Report(triggerData, aggregatedTotalRuntime, aggregatedTotalRuntimeWithBindings, blocked);
+        };
+    }
+
+    private ResponseDetails GetResponseDetails()
```

Comment:
This function could be converted to a switch expression if you felt so inclined 😄